lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -1918,6 +1927,83 @@ postTransactionFee ctx (ApiT wid) body = do
+    -> ArgGenChange s
+    -> ApiT WalletId
+    -> ApiConstructTransactionData n
+    -> Handler (ApiConstructTransaction n)
+constructTransaction ctx genChange (ApiT wid) body = do
+    let toAddressAmount (ApiPaymentAddresses content) =
+            addressAmountToTxOut <$> content
+        toAddressAmount (ApiPaymentAll _) = undefined -- this will be tackled when migration is supported
+    let md = body ^? #metadata . traverse . #getApiT
+    let pwd = coerce $ body ^. #passphrase . #getApiT
+    let mTTL = Nothing --this will be tackled when transaction validity is supported
+
+    (wdrl, mkRwdAcct) <-
+        mkRewardAccountBuilder @_ @s @_ @n ctx wid (body ^. #withdrawal)
+
+    ttl <- liftIO $ W.getTxExpiry ti mTTL
+    let txCtx = defaultTransactionCtx
+            { txWithdrawal = wdrl
+            , txMetadata = md
+            , txTimeToLive = ttl
+            --, txDelegationAction -- this will be tackled when delegations are supported
+            }
+
+    let transform = \s sel ->
+            W.assignChangeAddresses genChange sel s
+            & uncurry (W.selectionToUnsignedTx (txWithdrawal txCtx))
+
+    (apiSelection, fee, blob ) <- withWorkerCtx ctx wid liftE liftE $ \wrk -> do
+        w <- liftHandler $ W.readWalletUTxOIndex @_ @s @k wrk wid
+        let getFee = const (selectionDelta TokenBundle.getCoin)
+        (sel, sel', fee) <- case (body ^. #payments) of
+            Nothing -> do
+                utx <- liftHandler
+                    $ W.selectAssetsNoOutputs @_ @s @k wrk wid w txCtx (const Prelude.id)
+                (FeeEstimation estMin _) <- liftHandler $
+                    W.estimateFee $ W.selectAssetsNoOutputs @_ @s @k wrk wid w txCtx getFee
+                sel <- liftHandler $
+                    W.assignChangeAddressesWithoutDbUpdate wrk wid genChange utx
+                sel' <- liftHandler
+                    $ W.selectAssetsNoOutputs @_ @s @k wrk wid w txCtx transform
+                pure (sel, sel', estMin)
+
+            Just nonemptyPayments -> do
+                let outs = toAddressAmount nonemptyPayments
+                utx <- liftHandler
+                    $ W.selectAssets  @_ @s @k wrk w txCtx outs (const Prelude.id)
+                (FeeEstimation estMin _) <- liftHandler $ W.estimateFee $ W.selectAssets @_ @s @k wrk w txCtx outs getFee
+                sel <- liftHandler $
+                    W.assignChangeAddressesWithoutDbUpdate wrk wid genChange utx
+                sel' <- liftHandler
+                    $ W.selectAssetsNoOutputs @_ @s @k wrk wid w txCtx transform
+                pure (sel, sel', estMin)
+
+        blob <- liftHandler
+            $ W.constructUnsignedTransaction @_ @s @k wrk wid mkRwdAcct pwd txCtx sel
+
+        pure (mkApiCoinSelection [] Nothing md sel', fee, blob)
+
+    let txSerialized = ApiSerialisedTransaction (ApiBytesT $ convertToBase Base64 blob)
+    pure $ ApiConstructTransaction txSerialized apiSelection (Quantity $ fromIntegral fee)
```

Comment:
I think it would be nicer if there were a separate:
```haskell
mkApiConstructTransaction :: sel -> fee -> Either ErrMkTx ApiConstructTransaction
```

Currently `W.constructUnsignedTransaction` is getting the reward account XPrv, but only the XPub is necessary for the TxBody, right?

Reply:
indeed, in the end XPub is needed.